Add new inplace tag (#2388)
python-discord/bot: bot/resources/tags/in-place.md

**Out of Place** and **In Place**

> In programming, there are two types of operations: "out of place" operations creates a new object, leaving the original object unchanged. "in place" operations modifies the original object, without creating a new one. These operations return None explicitly.

Reviewer: ```suggestion
In programming, there are two types of operations: "out of place" operations create a new object, leaving the original object unchanged. "in place" operations modify the original object, without creating a new one. These operations return None explicitly.
```

A common example of these different concepts is seen in the use of the methods `list.sort()` and `sorted(...)`. Using `list.sort()` and attempting to access an element of the list will result in an error. 

```py
a_list = [3, 1, 2]
a_new_list = a_list.sort() # This will be None
print(a_new_list[1]) # This will error because it is NoneType and not a list

> a_list = [3, 1, 2]

Reviewer: It may be a bit clearer to use a new name here.

Author: For both or what exact list name

Author: Ah i see, resolved

sorted(a_list)
print(a_list[0]) # You may expect 1, but it will print 3
```

> 

Reviewer: This empty line can be removed, Discord's rendering auto-inserts one after code blocks.

> To avoid these errors and unexpected results, it is required to assign the result of `sorted(...)` to a new variable and use `list.sort()` method in the original list. This way, the original list will be sorted and the new list will be created with the sorted elements.

Reviewer: I'm not sure how I like "it is required" here, as well as the focus on lists. My proposal here would something along the lines of:

> To avoid these errors and unexpected results, you should either use an in-place operation (`sorted(...)`) and assign it to a variable or use an out-of-place operation (`list.sort()`) without assignment.

The wording can obviously vary, but I'd prefer to see something general like that.

Author: Seems good to me, I chose required because I couldn't really think of anything else in the moment.

Reviewer 2: I think variable names here should be more in-line with what the person was trying to do, so I would call them `unsorted_list` and `sorted_list` for the first example and `list_to_sort` in the second one.

I also think the way you show the unexpected behavior (with indexing) is overcomplicated, something like
```py
print(a_new_list)  # Outputs None. Where did the list go?
```
and
```py
print(outofplace_list)  # The list still isn't sorted. Why?
```
would make it easier to understand for beginners.

Author: @n0Oo0Oo0b Seems like a good idea, updated.

> **Out of Place** and **In Place**

Reviewer 3: You should add it as embed title according to this #2419

In programming, there are two types of operations: "out of place" operations create a new object, leaving the original object unchanged. "in place" operations modify the original object, without creating a new one. These operations return None explicitly.

A common example of these different concepts is seen in the use of the methods `list.sort()` and `sorted(...)`. Using `list.sort()` and attempting to access an element of the list will result in an error. 

```py
# WRONG:

unsorted_list = [3, 1, 2]

> sorted_list = inplace_list.sort() # This will be None

Reviewer: ```suggestion
sorted_list = unsorted_list.sort() # This will be None
```

Author: 👍, updated.

> sorted_list = unsorted_list.sort() # This will be None

Reviewer 3: [PEP 257](https://peps.python.org/pep-0008/#inline-comments) 
> An inline comment is a comment on the same line as a statement. Inline comments should be separated by at least two spaces from the statement. They should start with a # and a single space.

Should be changed to
```python
sorted_list = unsorted_list.sort()  # This will be None
```

---
embed:
    title: "Out of place and in place"
---

> In programming, there are two types of operations: "out of place" operations create a new object, leaving the original object unchanged. "in place" operations modify the original object, without creating a new one. These operations return None explicitly.

Reviewer: ```suggestion
In programming, there are two types of operations: "out of place" operations create a new object, leaving the original object unchanged. "in place" operations modify the original object without creating a new one, and return `None` explicitly.
```

**Out of Place** and **In Place**

In programming, there are two types of operations: "out of place" and "in place". An "out of place" operation creates a new object, leaving the original object unchanged. An "in place" operation modifies the original object, without creating a new one. These operations return None explicitly.

A common example of these different concepts is seen in the use of the methods `list.sort()` and sorted(...) in Python. Using `list.sort()` will modify the original list and return None, so attempting to access an element of the list after calling `sort()` will result in an error. 

For example, the following code will result in an error:

```py
a_list = [3, 1, 2]
a_new_list = a_list.sort() # This will be None

> print(a_new_list[1]) # This will error

Reviewer 4: ```suggestion
print(a_new_list[1]) # This will error because it is empty
```
Is this part necessary?

```

> On the other hand, using `sorted(...)` will return a new sorted list, leaving the original list unchanged. This means that if you expect the original list to be sorted, you will be disappointed with the result. For example, the following code will print 3 instead of 1:

Reviewer 4: ```suggestion
On the other hand, using the function `sorted(...)` will return a new sorted list, leaving the original list unchanged. This means that if you expect the original list to be sorted, you will be disappointed with the result. For example, the following code will print 3 instead of 1:
```

```py
a_list = [3, 1, 2]
sorted(a_list)
print(a_list[0]) # You may expect 1, but it will print 3
```

> To avoid these errors and unexpected results, it is required to assign the result of sorted(...) to a new variable and use list.sort() method in the original list. This way, the original list will be sorted and the new list will be created with the sorted elements.

Reviewer 4: ```suggestion
To avoid these errors and unexpected results, it is required to assign the result of `sorted(...)` to a new variable and use `list.sort()` method in the original list. This way, the original list will be sorted and the new list will be created with the sorted elements.
```
